contrib/NeonVM/global.c

```diff
@@ -0,0 +1,541 @@
+
+    for (int64_t i = fst; i < lst + 1; i++) {
+        cell_copyCell(&r->array[ri++], &array->array[i]);
+    }
+    pop(stack);
+
+    push(stack, r);
+}
+
+void array__splice(TStack *stack)
+{
+    BOOL last_from_end = top(stack)->boolean;   pop(stack);
+    Number last  = top(stack)->number;          pop(stack);
+    BOOL first_from_end  = top(stack)->boolean; pop(stack);
+    Number first = top(stack)->number;          pop(stack);
+    Cell *a = cell_fromCell(top(stack));        pop(stack);
+    Cell *b = cell_fromCell(top(stack));        pop(stack);
+    const Cell *array = a;
+
+    int64_t fst = bid128_to_int64_int(first);
+    int64_t lst = bid128_to_int64_int(last);
+    if (first_from_end) {
+        fst += array->array_size - 1;
+    }
+    if (fst < 0) {
+        fst = 0;
+    }
+    if (fst > (int64_t)array->array_size) {
+        fst = array->array_size;
+    }
+    if (last_from_end) {
+        lst += array->array_size - 1;
+    }
+    if (lst < -1) {
+        lst = -1;
+    }
+    if (lst >= (int64_t)array->array_size) {
+        lst = array->array_size - 1;
+    }
+
+    Cell *r = cell_createArrayCell(b->array_size + (array->array_size - (lst - fst)) - 1);
+    int64_t ai = 0;
+    for (int64_t i = 0; i < fst; i++) {
+        cell_copyCell(&r->array[ai++], &array->array[i]);
+    }
+    for (uint64_t i = 0; i < b->array_size; i++) {
+        cell_copyCell(&r->array[ai++], &b->array[i]);
+    }
+    for (uint64_t i = lst + 1; i < array->array_size; i++) {
+        cell_copyCell(&r->array[ai++], &array->array[i]);
+    }
+    cell_freeCell(b);
+    cell_freeCell(a);
+
+    push(stack, r);
+}
+
+void array__toBytes__number(TStack *stack)
+{
+    uint64_t x, i;
+    Cell *a = cell_fromCell(top(stack)); pop(stack);
+
+    Cell *r = cell_newCellType(cString);
+    r->string = string_createString(a->array_size);
+
+    for (x = 0, i = 0; x < a->array_size; x++) {
+        uint64_t b = bid128_to_uint64_int(a->array[x].number);
+        if (b >= 256) {
+            assert(b <= 256);
+            //throw RtlException(Exception_ByteOutOfRangeException, std::to_string(b));
+        }
+        r->string->data[i++] = (uint8_t)b;
+    }
+
+    cell_freeCell(a);
+    push(stack, r);
+}
+
+void array__toString__number(TStack *stack)
+{
+    uint64_t x;
+    // ToDo: Fix this to use proper string appends!!
+    Cell *a = top(stack);
+    char *s = malloc(STRING_BUFFER_SIZE);
+    // ToDo: Remove c strings
+    strcpy(s, "[");
+    for (x = 0; x < a->array_size; x++) {
+        if (strlen(s) > 1) {
+            strcat(s, ", ");
+        }
+        strcat(s, number_to_string(a->array[x].number));
+    }
+    strcat(s, "]");
+    pop(stack);
+    push(stack, cell_fromString(s, strlen(s)));
+    free(s);
+}
+
+
+
+void boolean__toString(TStack *stack)
+{
+    BOOL v = top(stack)->boolean; pop(stack);
+    push(stack, cell_fromString(v ? "TRUE" : "FALSE", -1));
+}
+
+
+
+void bytes__decodeToString(TStack *stack)
+{
+    stack = stack;
```

Comment:
What's this for? Is this just to avoid a compiler warning that 'stack' is otherwise unused?

Reply:
Commented.